snapcraft/cli/_channel_map.py

```diff
@@ -0,0 +1,267 @@
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import os
+from collections import OrderedDict
+from typing import List, NamedTuple, Optional
+
+from tabulate import tabulate
+
+from snapcraft.storeapi.v2.channel_map import (
+    ChannelMap,
+    MappedChannel,
+    Revision,
+    SnapChannel,
+)
+
+
+_HINTS_T = NamedTuple(
+    "_HINTS_T",
+    [
+        ("CLOSED", str),
+        ("FOLLOWING", str),
+        ("NO_PROGRESS", str),
+        ("PROGRESSING_TO", str),
+    ],
+)
+_HINTS = _HINTS_T("-", "↑", "-", "→")
+
+
+def _get_channel_hint(*, channel_map, fallback: str, architecture: str) -> str:
+    tick = _HINTS.CLOSED
+    for c in channel_map:
+        if c.channel == fallback and c.architecture == architecture:
+            tick = _HINTS.FOLLOWING
+            break
+    else:
+        if fallback is None:
+            tick = _HINTS.CLOSED
+    return tick
+
+
+def _get_channel_order(
+    snap_channels, tracks: Optional[List[str]] = None
+) -> OrderedDict:
+    channel_order: OrderedDict = OrderedDict()
+
+    if tracks is not None:
+        snap_channels = [s for s in snap_channels if s.track in tracks]
+
+    for snap_channel in snap_channels:
+        if snap_channel.track not in channel_order:
+            channel_order[snap_channel.track] = list()
+        if snap_channel.fallback is None:
+            channel_order[snap_channel.track].append(snap_channel.name)
+        else:
+            try:
+                channel_order[snap_channel.track].insert(
+                    channel_order[snap_channel.track].index(snap_channel.fallback) + 1,
+                    snap_channel.name,
+                )
+            except ValueError:
+                channel_order[snap_channel.track].append(snap_channel.name)
+
+    return channel_order
+
+
+def _get_channel_line(
+    *,
+    mapped_channel: Optional[MappedChannel],
+    revision: Optional[Revision],
+    channel_info: SnapChannel,
+    hint: str,
+    progress_string: str,
+) -> List[str]:
+    version_string = hint
+    revision_string = hint
+    expiration_date_string = ""
+    channel_string = channel_info.risk
+
+    if revision is not None:
+        version_string = revision.version
+        revision_string = f"{revision.revision}"
+
+    if mapped_channel is not None:
+        if channel_info.branch is None and mapped_channel.progressive.percentage:
+            channel_string = ""
+        elif channel_info.branch is not None:
+            channel_string = f"{channel_info.risk}/{channel_info.branch}"
+            # This value can be None.
```

Comment:
Can probably drop the comment if check is:
`if mapped_channel.expiration_date is not None`

Reply:
done